beest/cmd/driver/pytest.go

```diff
@@ -0,0 +1,126 @@
+package driver
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"os/exec"
+	"strings"
+	"syscall"
+)
+
+const (
+	DefaultPyTestBinary      = "py.test"
+	DefaultPyTestWatchBinary = "ptw"
+
+	SshIdentityFileFlag  = "--ssh-identity-file"
+	AnsibleInventoryFlag = "--ansible-inventory"
+	HostsFlag            = "--hosts"
+	ConnectionFlag       = "--connection"
+
+	ExpressionFlag = "-k"
+)
+
+type PytestContext interface {
+	ConnectionContext
+	TestsPath() string
+	Hostnames() []string
+}
+
+func PyTestRun(ctx PytestContext, watch bool, selection string) error {
+	cmd := buildPyTestCmd(ctx, watch, selection)
+	log.Printf("Running Pytest cmd: %s", cmd)
+	return runPyTestCmd(context.Background(), cmd)
+}
+
+func buildPyTestCmd(ctx PytestContext, watch bool, selection string) *exec.Cmd {
+	var args []string
+
+	if watch {
+		args = append(args, "--")
+	}
+
+	args = append(args, "-vs") //TODO debug
+	args = append(args, "-rap")
+
+	args = append(args, fmt.Sprintf("%s=%s", SshIdentityFileFlag, ctx.PrivateKey()))
+	args = append(args, fmt.Sprintf("%s=%s", ConnectionFlag, "ansible")) //Ansible module is only available with ansible connection backend
+	args = append(args, fmt.Sprintf("%s=%s", AnsibleInventoryFlag, ctx.Inventory()))
+
+	if len(ctx.Hostnames()) > 0 {
+		var hosts []string
+		for _, hostName := range ctx.Hostnames() {
+			hosts = append(hosts, fmt.Sprintf("ansible://%s", hostName))
+		}
+		args = append(args, fmt.Sprintf("%s=%s", HostsFlag, strings.Join(hosts, ",")))
+	}
+
+	if selection != "" {
+		args = append(args, fmt.Sprintf("%s %s", ExpressionFlag, selection))
+	}
+
+	var cmd *exec.Cmd
+	if watch {
+		cmd = exec.Command(DefaultPyTestWatchBinary, args...)
+	} else {
+		cmd = exec.Command(DefaultPyTestBinary, args...)
+	}
+
+	// TODO env variables
+	//cmd.Env = tf.buildEnv(mergeEnv)
+	cmd.Dir = ctx.TestsPath()
+
+	return cmd
+}
+
+func runPyTestCmd(ctx context.Context, cmd *exec.Cmd) error {
+	var errBuf strings.Builder
+
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = io.MultiWriter(os.Stderr, &errBuf)
+
+	cmd.SysProcAttr = &syscall.SysProcAttr{
```

Comment:
Interestingly Windows has a totally different model for `syscall.SysProcAttr`. I doubt anyone will run it on windows I just noticed it on my machine now, but it's probably good to add a `!windows` build flag.

Reply:
I guess that also means you have to implement a `pytest_windows.go` but just throwing Unimplemented error for now should be fine

Reply:
agreed